### DISCUSSION TITLE
Is there a way to create an Surface Shader?

### DISCUSSION THREAD

**Discussion by cesarl94** [jnsmalm/pixi3d#206] [General]
I'm noticed that you can create your own shader like this example: https://codesandbox.io/s/pixi3d-custom-material-4tsyb
with your custom vertex and your custom fragment, but when you do that, you lose the benefits of a standard material.
In Unreal Engine, when you create a Material, you put the logic in a BP that represents your parameters of the Standard Material node that knows what to do with the lights, and everything in the world.
In Unity, is similar but in code; they have a "surf" function that is the same place where you put the logic for calculate your parameters and then apply the Standard Material magic.

Is there something similar in pixi3d? Or I need implement my custom fragment?

**Comment by jnsmalm** [comment]
This is something I wanted to support but couldn't figure out how to do it in a good way. Yes, you need to create your own custom vertex/fragment shader.

**Reply by cesarl94** [reply]
I'm studying the project trying to implement it. If that's okay with you, I would like to collaborate on the project with that

**Reply by jnsmalm** [reply]
Sure, If you have any good ideas I would be happy to hear about them.

**Reply by cesarl94** [reply]
Good! I will to write that between today and tomorrow and I will answer you here with that I have in mind

**Reply by cesarl94** [reply]
Well, to start, I propose creating a class called "StandardSurfaceMaterial," which, unlike StandardMaterial, will not have predefined textures or material parameters. Instead, it will have various functions to receive different classes of parameters along with an identifier string. Among these: textures, vectors, colors (which are vectors after all), scalars, and integers, and it will store them internally.

On the other hand, this class will receive shader code pieces, in which you can assume you are "inside the fragment," and you can:

1) Overwrite a series of variables that represent the parameters of any surface shader (baseColor, metallic, roughness, emissive, normal, etc.). If you don't use them, it's okay because they will have default values.
2) Read the parameters entered previously by the string provided at the time of loading, modify them, and pass data around.
3) Use the rest of the shader utilities, such as obtaining the UV coordinates, the global position of the pixel, getting the vertexColor, etc.
Once we have all this, the class will be responsible for including those provided parameters as uniforms and embedding that piece of code into the fragment template.

Initially, this will require writing another new fragment, reusing the logic of the standard material we already have but adapting it so that this can work, and then we can merge them to avoid repeated code.

This has to come with documentation and examples that I am willing to create.

**Reply by jnsmalm** [reply]
Sounds like good ideas. All this can be created outside of Pixi3D, I think that will be easier for you (take a look at the code for StandardMaterial to get started). Although I don't think this feature will be easy to get right, you can have a go at it if you want :) If it turns out good we can figure out a way to add it to Pixi3D package.

**Reply by cesarl94** [reply]
I will try. When I get it, I will show you the results
